File: problems-extend-adapt/2301-2400/2345_counting-uncovered-summits/solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {

    public int uncoveredSummits(int[][] peaks) {
        // Mountain (x, y) contains peak (a, b) exactly when |a - x| <= y - b:
        // the peak sits inside or on the slopes. Sorting by x ascending
        // (ties by y descending) puts every potential coverer no later, so
        // a monotonic stack settles everything in one pass. Duplicated
        // peaks are invisible but still hide others, so they stay on the
        // stack for their covering effect and are only excluded from the
        // final count.
        Arrays.sort(peaks, (p, q) -> {
            if (p[0] != q[0]) {
                return Integer.compare(p[0], q[0]);
            }
            return Integer.compare(q[1], p[1]);
        });
        List<int[]> stack = new ArrayList<>(); // {x, y, counted}
        int i = 0;
        while (i < peaks.length) {
            int j = i; // run-length encode equal peaks to detect duplicates
            while (j < peaks.length && peaks[j][0] == peaks[i][0] && peaks[j][1] == peaks[i][1]) {
                ++j;
            }
            boolean duplicated = j - i > 1;
            int x = peaks[i][0],
                y = peaks[i][1];
            while (
                !stack.isEmpty() && Math.abs(stack.get(stack.size() - 1)[0] - x) <= y - stack.get(stack.size() - 1)[1]
            ) {
                stack.remove(stack.size() - 1);
            }
            boolean covered =
                !stack.isEmpty() && Math.abs(x - stack.get(stack.size() - 1)[0]) <= stack.get(stack.size() - 1)[1] - y;
            if (!covered) {
                stack.add(new int[] { x, y, duplicated ? 0 : 1 });
            }
            i = j;
        }
        int visible = 0;
        for (int[] entry : stack) {
            visible += entry[2];
        }
        return visible;
    }
}

File: problems-extend-adapt/2301-2400/2345_counting-uncovered-summits/solution.go
import "sort"

func uncoveredSummits(peaks [][]int) int {
	// Mountain (x, y) contains peak (a, b) exactly when |a - x| <= y - b:
	// the peak sits inside or on the slopes. Sorting by x ascending (ties
	// by y descending) puts every potential coverer no later, so a
	// monotonic stack settles everything in one pass. Duplicated peaks are
	// invisible but still hide others, so they stay on the stack for their
	// covering effect and are only excluded from the final count.
	sort.Slice(peaks, func(i, j int) bool {
		if peaks[i][0] != peaks[j][0] {
			return peaks[i][0] < peaks[j][0]
		}
		return peaks[i][1] > peaks[j][1]
	})
	type entry struct{ x, y, counted int }
	var stack []entry
	for i := 0; i < len(peaks); {
		j := i
		for j < len(peaks) && peaks[j][0] == peaks[i][0] && peaks[j][1] == peaks[i][1] {
			j++
		}
		duplicated := j-i > 1
		x, y := peaks[i][0], peaks[i][1]
		for len(stack) > 0 && abs(stack[len(stack)-1].x-x) <= y-stack[len(stack)-1].y {
			stack = stack[:len(stack)-1]
		}
		covered := false
		if len(stack) > 0 {
			top := stack[len(stack)-1]
			covered = abs(x-top.x) <= top.y-y
		}
		counted := 0
		if !duplicated {
			counted = 1
		}
		if !covered {
			stack = append(stack, entry{x, y, counted})
		}
		i = j
	}
	visible := 0
	for _, e := range stack {
		visible += e.counted
	}
	return visible
}

func abs(v int) int {
	if v < 0 {
		return -v
	}
	return v
}
